### PR TITLE
feat(ec2): Support explicit Subnet selection

### LINKED ISSUE #4604: Attach Subnet to AutoScaleGroup
AWS AutoScaling Groups is a tool to provision EC2 instances for various workload patterns. AWS CDK allows developers to specify `vpcSubnets` subnets selection (meaning existing one) but do not allow to create Subnet and ASG within same stack.

### Use Case

This issue is continuation of use-cases defined in #4586. There is a need to provision backend infrastructure within own subnet. There is a stack that deploys extension to existing VPC.

```typescript 
const vpc = ec2.Vpc.fromLookup(this, 'Vpc', {/* ... */})

const subnet1 = new ec2.PrivateSubnet(this, 'BE1', {
  availabilityZone: vpc.availabilityZones[0],
  // ...
})

const subnet2 = new ec2.PrivateSubnet(this, 'BE2', {
  availabilityZone: vpc.availabilityZones[1],
  // ...
})

const nodes = new asg.AutoScalingGroup(this, 'Nodes', {
   vpcSubnets: {/* impossible to pass reference to subnet1 and subnet2 */}
   // ...
})
```

It is **impossible** to associate ASG to subnets. There is only one ugly work around - you have to declare the subnets in own stack and then import to ASG using `subnetGroupName`. However, workaround defeats purpose of layers design (see  #4586).

### Proposed Solution

**Option 1: brute-force design**
Easy to design, implement, impacts only AutoScalingGroup. I do not like it. 

Enhance the property type `vpcSubnets: SubnetSelection | string[]`. Then modify following code to either pick subnets id as-is or query data from vpc.

https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts#L488

**Option 2: support explicit subnet definition with SubnetSelection**

Give possibility to explicitly declare subnets ID as part of the subnet selection query

```typescript
interface SubnetSelection {
   /**
   * Select the subnet group with the given ids
   */
   readonly subnetIds?: ISubnet[];
}
``` 

then support this attribute at `VPC.selectSubnets(...)`, which acts roughly as transform function which maps `ISubnet` to `SelectedSubnets`.

I'd recommend a second approach, it allows to support subnet assignment to other classes as well. Not only too ASG.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by fogfish**
closes #4604 

__WHY__
AWS CDK allows developers to attach subnets to other resource (e.g. ASG)
using Subnets selection feature. It implies re-use of existing subnets.
It is not possible to create subnet and attach it to ASG in same stack.

__WHAT__
The Subnet selection feature is expanded with ability to explicitly
define desired subnets. The subnet selection essentially becomes an
identity funnction.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 1]
```diff
@@ -1464,7 +1475,15 @@ function reifySelectionDefaults(placement: SubnetSelection): SubnetSelection {
     throw new Error(`Only one of 'subnetType' and 'subnetGroupName' can be supplied`);
   }
 
-  if (placement.subnetType === undefined && placement.subnetGroupName === undefined) {
+  if (placement.subnets !== undefined && placement.subnetType !== undefined) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts]: Can you make this `if` compare all 3 parameters at the same time and produce an appropriate error message?

Suggested code:

```ts
const possibleKeys: Array<keyof SubnetSelection> = ['subnets', 'subnetType', 'subnetGroupName'];
const provided = keys.filter(key => placement[key] !== undefined);

if (provided.length > 1) {
  throw new Error(`'${provided}' specified together in subnet selection, specify only one`);
}
```

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 2]
```diff
@@ -205,6 +205,14 @@ export interface SubnetSelection {
    * @default false
    */
   readonly onePerAz?: boolean;
+
+  /**
+   * Explicit Subnet selection.
+   *
+   * The subnet selection is an identity function that binds
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts]: I feel this is too technical for most of the intended audience.

How about:

```
Explicitly select individual subnets

Use this if you don't want to automatically use all subnets in
a group, but have a need to control selection down to
individual subnets.

Cannot be specified together with `subnetType` or `subnetGroupName`.

@default - Use all subnets in a selected group (all private subnets by default)
```

**Comment by rix0rrr** [comment]
Thanks for contributing this. Some small changes please and then it can go right in.

**Comment by fogfish** [comment]
@rix0rrr Thank you for comments! They are in.

**Pull request merged**